distribution/tools/entitlement-agent/src/main/java/org/elasticsearch/entitlement/agent/EntitlementAgent.java

```diff
@@ -9,13 +9,51 @@
 
 package org.elasticsearch.entitlement.agent;
 
-import org.elasticsearch.entitlement.runtime.api.EntitlementChecks;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.instrumentation.Instrumenter;
 
+import java.io.File;
+import java.io.IOException;
 import java.lang.instrument.Instrumentation;
+import java.util.List;
+import java.util.jar.JarFile;
+
+import static java.util.stream.Collectors.toSet;
 
 public class EntitlementAgent {
 
     public static void premain(String agentArgs, Instrumentation inst) throws Exception {
-        EntitlementChecks.setAgentBooted();
+        // Add the trampoline library (the one with the entitlement checking interface) to the bootstrap classpath.
+        // We can't actually reference the classes here for real before this point because they won't resolve.
+        var jarsString = System.getProperty("entitlements.trampolineJars");
+        if (jarsString != null) {
+            addJarsToBootstrapClassLoader(inst, jarsString);
+        }
+
+        ConfigurationScanner.ScanResults config = ConfigurationScanner.scan(List.of(EntitlementChecks.class));
+
+        inst.addTransformer(
+            new Transformer(
+                new Instrumenter("", config.methodMap()),
+                config.classesToInstrument().stream().map(EntitlementAgent::internalName).collect(toSet())
+            ),
+            true
+        );
+        inst.retransformClasses(config.classesToInstrument().toArray(new Class<?>[0]));
```

Comment:
The trampoline code already references them, so if that's a problem, the horse has already left the barn by this point.

If we want to avoid loading certain classes that need instrumentation, we might need a different approach in the trampoline code.

Reply:
How does the trampoline cause these classes to be loaded?

Reply:
The trampoline layer includes the `EntitlementChecks` interface which has all the methods called from the instrumented code. For each of those methods, the second argument is the target class. That means `EntitlementChecks` references all the target classes.

The agent loads `EntitlementChecks` to scan it to discover what to instrument, and will use reflection on those classes to look up the appropriate methods. I think it's pretty likely that this will cause the agent to load all the target classes (though I haven't confirmed this).

Reply:
> What happens if a class hasn't been loaded yet? Will it force loading here, or skip the class? I'm wondering if we need to filter this down to a set of classes like System that we know will need to be reloaded.

`retransformClasses` will process again classes already loaded. It is to bridge the gap between the moment the JVM started and we got to this point: things like String and File will always be already loaded at this point.

It will _not_ force loading any class; for that, the Instrument API will use the transformer delcared a few lines above at load time and/or at redefinition time.